### PR TITLE
Enable dpi awareness windows

### PR DESCRIPTION
**Pull request by GioPat**
This PR is enabling DPI awareness for windows target and thus solving issue #99.
This is the result on my system:
**Before:**
image

**After:**
image

Thanks to @sopium and @parkovski for the insights (review and ideas).

### CODE REVIEW

**Comment by zxey** [comment]
Thanks raising and issue and creating a PR. Is this ready to be merged?

**Comment by sopium** [comment]
I don't think so. Edge code does not compile yet.

@GioPat you can test it with `--features=edge`.

File: webview-sys/webview_mshtml.c [hunk 1]
```diff
@@ -110,6 +110,9 @@ typedef struct {
   _IServiceProviderEx provider;
 } _IOleClientSiteEx;
 
+typedef BOOL (*SetThreadDpiAwareness)(DPI_AWARENESS_CONTEXT);
```
Comment (sopium) [inline webview-sys/webview_mshtml.c]: As @parkovski have mentioned, the calling convention is significant on x86. (But does not matter on x86_64.)

```suggestion
typedef BOOL (WINAPI *SetThreadDpiAwarenessContext)(DPI_AWARENESS_CONTEXT);
```

File: webview-sys/webview_mshtml.c [hunk 2]
```diff
@@ -110,6 +110,9 @@ typedef struct {
   _IServiceProviderEx provider;
 } _IOleClientSiteEx;
 
+typedef BOOL (*SetThreadDpiAwareness)(DPI_AWARENESS_CONTEXT);
+typedef BOOL (*SetProcessDpiAwareness)(int);
```
Comment (sopium) [inline webview-sys/webview_mshtml.c]:
```suggestion
typedef BOOL (WINAPI* SetProcessDpiAwareness)(int);
```

**Comment by sopium** [comment]
> 
> It's a bit messy to get DPI value. I found [this method](https://docs.microsoft.com/en-gb/windows/win32/api/shellscalingapi/nf-shellscalingapi-getdpiformonitor?redirectedfrom=MSDN) that retrieves dpiX and dpiY but requires shcore.dll to be loaded.

I think you can use `GetDpiForWindow(GetDesktopWindow())` (call `GetDpiForWindow` with `GetProcAddress`, because it's only available on Windows 10), and fallback to [`GetDeviceCaps`](https://docs.microsoft.com/en-us/windows/win32/directwrite/how-to-ensure-that-your-application-displays-properly-on-high-dpi-displays#gdi).

File: webview-sys/webview_edge.cpp [hunk 3]
```diff
@@ -238,6 +270,21 @@ LRESULT CALLBACK WebviewWndProc(HWND hwnd, UINT msg, WPARAM wp, LPARAM lp)
...
         w->resize();
         break;
+    case WM_DPICHANGED:
+        auto rect = reinterpret_cast<LPRECT>(lp);
+        auto x = rect->left;
+        auto y = rect->top;
+        auto w = rect->right - x;
+        auto h = rect->bottom - y;
+        SetWindowPos(hwnd, nullptr, x, y, w, h, SWP_NOZORDER);
+        auto monitor = MonitorFromWindow(hwnd, MONITOR_DEFAULTTONEAREST);
```
Comment (parkovski) [inline webview-sys/webview_edge.cpp]: If you aren't caching the scale/resolution then just end this block at the `SetWindowPos` call.

File: webview-sys/webview_edge.cpp [hunk 4]
```diff
@@ -238,6 +270,21 @@ LRESULT CALLBACK WebviewWndProc(HWND hwnd, UINT msg, WPARAM wp, LPARAM lp)
...
+        MONITORINFOEXW mi{};
+        mi.cbSize = sizeof(MONITORINFOEXW);
+        GetMonitorInfoW(monitor, &mi);
+        auto scale = LOWORD(wp) / (float)USER_DEFAULT_SCREEN_DPI;
+        auto xres = mi.rcMonitor.right - mi.rcMonitor.left;
+        auto yres = mi.rcMonitor.bottom - mi.rcMonitor.top;
+        break;
```
Comment (parkovski) [inline webview-sys/webview_edge.cpp]: This `break` should be `return 0` (see https://docs.microsoft.com/en-us/windows/win32/hidpi/wm-dpichanged).

File: webview-sys/webview_mshtml.c [hunk 5]
```diff
@@ -786,6 +823,23 @@ static LRESULT CALLBACK wndproc(HWND hwnd, UINT uMsg, WPARAM wParam,
...
+    auto h = rect.bottom - y;
+    SetWindowPos(hwnd, NULL, x, y, w, h, SWP_NOZORDER);
+    auto monitor = MonitorFromWindow(hwnd, MONITOR_DEFAULTTONEAREST);
+    MONITORINFOEXW mi;
+    mi.cbSize = sizeof(MONITORINFOEXW);
+    GetMonitorInfoW(monitor, &mi);
+    auto scale = LOWORD(wParam) / (float)USER_DEFAULT_SCREEN_DPI;
+    auto xres = mi.rcMonitor.right - mi.rcMonitor.left;
+    auto yres = mi.rcMonitor.bottom - mi.rcMonitor.top;
+    return TRUE;
```
Comment (parkovski) [inline webview-sys/webview_mshtml.c]: MSDN says you should return 0 here.

**Comment by sopium** [comment]
Does anyone know how to make the mshtml web browser control properly hande DPI *changes*? When I change DPI the document remains the old size.

If we can't figure this out maybe we should use “system DPI awareness” for mshtml. When DPI changes things will be correctly scaled (but blurry).

**Comment by zxey** [comment]
Sorry I couldn't get to you quickly, if you still want to finish this, I am okay with using "system DPI awarenes" for mshtml.

**Comment by GioPat** [comment]
> Sorry I couldn't get to you quickly, if you still want to finish this, I am okay with using "system DPI awarenes" for mshtml.

Hello zxey, I can try to solve the open issues and proceed with the merge! :)

**Comment by GioPat** [comment]
I applied all the comments, still don't get why the travis build is failing. On my windows OS is compiling good also with --feature edge enabled.

**Comment by sopium** [comment]
I think the errors in CI are specific to mingw(gcc). Maybe it's an older gcc that simply does not have those definitions. Or maybe we need to define `WINVER`.

I don't have a windows box now, but I can reproduce the same errors with [`cross`](https://github.com/rust-embedded/cross): `cross check --target=x86_64-pc-windows-gnu`.

We can replicate those definitions when they are undefined:

```C
#ifndef DPI_AWARENESS_CONTEXT_PER_MONITOR_AWARE_V2
DECLARE_HANDLE(DPI_AWARENESS_CONTEXT);
#define DPI_AWARENESS_CONTEXT_PER_MONITOR_AWARE_V2 ((DPI_AWARENESS_CONTEXT)-4)
#endif
#ifndef WM_DPICHANGED
#define WM_DPICHANGED 0x02e0
#endif
#ifndef USER_DEFAULT_SCREEN_DPI
#define USER_DEFAULT_SCREEN_DPI 96
#endif
```

P.S. I think in CI we should test with both msvc and gnu on windows if we intend to support both. And test the edge feature (with msvc, as gnu is not supported it seems).

**Comment by zxey** [comment]
Thanks for contributing this, I merged yours and sopium's changes as part of #134.

**Pull request closed without merging**